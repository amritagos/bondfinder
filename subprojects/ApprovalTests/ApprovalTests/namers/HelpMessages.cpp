#include "ApprovalTests/namers/HelpMessages.h"
#include "ApprovalTests/utilities/StringUtils.h"

#include <sstream>

namespace ApprovalTests
{

    std::string HelpMessages::getMisconfiguredBuildHelp(const std::string& fileName)
    {
        std::string helpMessage = R"(* Welcome to Approval Tests.
*
* There seems to be a problem with your build configuration.
* We cannot find the test source file at:
*   [fileName]
*
* For details on how to fix this, please visit:
* https://github.com/approvals/ApprovalTests.cpp/blob/master/doc/TroubleshootingMisconfiguredBuild.md
*
* For advanced users only:
* If you believe you have reached this message in error, you can bypass
* the check by calling ApprovalTestNamer::setCheckBuildConfig(false);
)";
        return StringUtils::replaceAll(
            topAndTailHelpMessage(helpMessage), "[fileName]", fileName);
    }

    std::string HelpMessages::getMisconfiguredMainHelp()
    {
        std::string helpMessage = R"(* Welcome to Approval Tests.
*
* You have forgotten to configure your test framework for Approval Tests.
*
* To do this in Catch, add the following to your main.cpp:
*
*     #define APPROVALS_CATCH
*     #include "ApprovalTests.hpp"
*
* To do this in Google Test, add the following to your main.cpp:
*
*     #define APPROVALS_GOOGLETEST
*     #include "ApprovalTests.hpp"
*
* To do this in doctest, add the following to your main.cpp:
*
*     #define APPROVALS_DOCTEST
*     #include "ApprovalTests.hpp"
*
* To do this in Boost.Test, add the following to your main.cpp:
*
*     #define APPROVALS_BOOSTTEST
*     #include "ApprovalTests.hpp"
*
* To do this in CppUTest, add the following to your main.cpp:
*
*     #define APPROVALS_CPPUTEST
*     #include "ApprovalTests.hpp"
*
* To do this in [Boost].UT, add the following to your main.cpp:
*
*     #define APPROVALS_UT
*     #include "ApprovalTests.hpp"
*
* For more information, please visit:
* https://github.com/approvals/ApprovalTests.cpp/blob/master/doc/TroubleshootingMisconfiguredMain.md
)";
        return topAndTailHelpMessage(helpMessage);
    }

    std::string HelpMessages::getUnconfiguredRootDirectory()
    {
        std::string helpMessage = R"(* Hello from Approval Tests.
*
* It looks like your program is calling some code that requires knowledge of
* the location of the main() in your source tree.
*
* To do this, add the following to your main.cpp file:
*
*     APPROVAL_TESTS_REGISTER_MAIN_DIRECTORY
*
* Currently, this is only required if you are using TemplatedCustomNamer's
* {RelativeTestSourceDirectory}.
)";
        return topAndTailHelpMessage(helpMessage);
    }

    std::string
    HelpMessages::getUnknownEnvVarReporterHelp(const std::string& envVarName,
                                               const std::string& selected,
                                               const std::vector<std::string>& knowns)
    {
        std::string helpMessage =
            R"(* The environment variable [envVarName] contains the value
* [selected]
*
* This reporter is not recognised.
*
* Please unset the environment value, or change it to refer to one of the
* known reporters:
*
[known]*
* For more information, see:
* https://github.com/approvals/ApprovalTests.cpp/blob/master/doc/how_tos/SelectReporterWithEnvironmentVariable.md
)";

        return envVarErrorMessage(envVarName, selected, knowns, helpMessage);
    }

    std::string
    HelpMessages::getInvalidEnvVarReporterHelp(const std::string& envVarName,
                                               const std::string& selected,
                                               const std::vector<std::string>& knowns)
    {
        std::string helpMessage =
            R"(* The environment variable [envVarName] contains the value
* [selected]
*
* This reporter is recognised, but cannot be found on this machine.
*
* Please unset the environment value, or change it to refer to a working
* reporter:
*
[known]*
* For more information, see:
* https://github.com/approvals/ApprovalTests.cpp/blob/master/doc/how_tos/SelectReporterWithEnvironmentVariable.md
)";

        return envVarErrorMessage(envVarName, selected, knowns, helpMessage);
    }

    std::string HelpMessages::envVarErrorMessage(const std::string& envVarName,
                                                 const std::string& selected,
                                                 const std::vector<std::string>& knowns,
                                                 std::string& helpMessage)
    {
        std::stringstream ss;
        for (auto& known : knowns)
        {
            ss << "* " << known << '\n';
        }
        helpMessage = StringUtils::replaceAll(helpMessage, "[selected]", selected);
        helpMessage = StringUtils::replaceAll(helpMessage, "[envVarName]", envVarName);
        return topAndTailHelpMessage(
            StringUtils::replaceAll(helpMessage, "[known]", ss.str()));
    }

    std::string HelpMessages::topAndTailHelpMessage(const std::string& message)
    {
        const std::string lineBreak =
            "**************************************************************"
            "***************";
        const std::string lineBuffer =
            "*                                                             "
            "              *\n";
        return lineBreak + '\n' + lineBuffer + message + lineBuffer + lineBreak;
    }
}
