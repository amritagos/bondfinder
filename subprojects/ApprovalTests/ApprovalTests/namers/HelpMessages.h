#pragma once

#include <string>
#include <vector>

namespace ApprovalTests
{
    class HelpMessages
    {
    public:
        static std::string getMisconfiguredBuildHelp(const std::string& fileName);

        static std::string getMisconfiguredMainHelp();

        static std::string getUnconfiguredRootDirectory();

        static std::string
        getUnknownEnvVarReporterHelp(const std::string& envVarName,
                                     const std::string& selected,
                                     const std::vector<std::string>& knowns);
        static std::string
        getInvalidEnvVarReporterHelp(const std::string& envVarName,
                                     const std::string& selected,
                                     const std::vector<std::string>& knowns);

        static std::string envVarErrorMessage(const std::string& envVarName,
                                              const std::string& selected,
                                              const std::vector<std::string>& knowns,
                                              std::string& helpMessage);

        static std::string topAndTailHelpMessage(const std::string& message);
    };
}
