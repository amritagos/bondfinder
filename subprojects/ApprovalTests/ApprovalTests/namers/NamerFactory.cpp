#include "ApprovalTests/namers/NamerFactory.h"

namespace ApprovalTests
{
    SectionNameDisposer
    NamerFactory::appendToOutputFilename(const std::string& sectionName)
    {
        return SectionNameDisposer(ApprovalTestNamer::currentTest(), sectionName);
    }
}
