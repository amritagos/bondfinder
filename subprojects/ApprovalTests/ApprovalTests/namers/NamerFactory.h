#pragma once

#include "ApprovalTests/namers/SectionNameDisposer.h"

#include <string>

namespace ApprovalTests
{
    struct NamerFactory
    {
        static SectionNameDisposer appendToOutputFilename(const std::string& sectionName);
    };
}
