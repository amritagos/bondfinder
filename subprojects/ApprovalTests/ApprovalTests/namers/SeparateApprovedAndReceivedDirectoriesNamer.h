#pragma once

#include "ApprovalTests/namers/ApprovalTestNamer.h"
#include "ApprovalTests/Approvals.h"
#include "ApprovalTests/namers/TemplatedCustomNamer.h"

namespace ApprovalTests
{
    class SeparateApprovedAndReceivedDirectoriesNamer : public TemplatedCustomNamer
    {
    public:
        SeparateApprovedAndReceivedDirectoriesNamer();

        virtual ~SeparateApprovedAndReceivedDirectoriesNamer() override = default;

        static DefaultNamerDisposer useAsDefaultNamer();
    };
}
