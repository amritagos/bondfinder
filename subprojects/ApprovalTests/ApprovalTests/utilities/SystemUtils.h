#pragma once

#include "ApprovalTests/utilities/WinMinGWUtils.h"
#ifdef _WIN32
// ReSharper disable once CppUnusedIncludeDirective
#include <io.h>
#include <direct.h>
#else
// ReSharper disable once CppUnusedIncludeDirective
#include <unistd.h>
#endif

#include "ApprovalTests/utilities/FileUtils.h"
#include "ApprovalTests/utilities/StringUtils.h"
#include "ApprovalTests/utilities/Macros.h"

#include <stdexcept>

namespace ApprovalTests
{
    class SystemUtils
    {
    public:
        static bool isWindowsOs();

        static bool isCygwin();

        static bool isMacOs();

        static std::string getDirectorySeparator();

        // Properly cases the filename, but not the directories, on Windows.
        static std::string checkFilenameCase(const std::string& fullPath);

        static std::string safeGetEnvForWindows(char const* name);

        static std::string safeGetEnvForNonWindows(char const* name);

        //! Return the value of the environment variable, or an empty string if the variable is not set.
        static std::string safeGetEnv(char const* name);

        static std::string getMachineName();

        static void makeDirectoryForWindows(const std::string& directory);

        static void makeDirectoryForNonWindows(const std::string& directory);

        static void makeDirectory(const std::string& directory);

        static void ensureDirectoryExists(const std::string& fullDirectoryPath);

        static void ensureParentDirectoryExists(const std::string& fullFilePath);

        static void runSystemCommandOrThrow(const std::string& command,
                                            bool allowNonZeroExitCodes = false);
    };
}
