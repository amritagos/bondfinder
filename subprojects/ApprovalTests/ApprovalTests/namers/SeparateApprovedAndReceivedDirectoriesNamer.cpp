#include "ApprovalTests/namers/SeparateApprovedAndReceivedDirectoriesNamer.h"
#include "ApprovalTests/utilities/SystemUtils.h"

namespace ApprovalTests
{
    std::string separateDirectoryPath()
    {
        // clang-format off
        // begin-snippet: separate_approved_and_received_directory_names
        auto path = "{TestSourceDirectory}/{ApprovalsSubdirectory}/{ApprovedOrReceived}/{TestFileName}.{TestCaseName}.{FileExtension}";
        // end-snippet
        // clang-format on
        return path;
    }

    SeparateApprovedAndReceivedDirectoriesNamer::
        SeparateApprovedAndReceivedDirectoriesNamer()
        : TemplatedCustomNamer(separateDirectoryPath())
    {
    }

    DefaultNamerDisposer SeparateApprovedAndReceivedDirectoriesNamer::useAsDefaultNamer()
    {
        return Approvals::useAsDefaultNamer([]() {
            return std::make_shared<SeparateApprovedAndReceivedDirectoriesNamer>();
        });
    }
}
