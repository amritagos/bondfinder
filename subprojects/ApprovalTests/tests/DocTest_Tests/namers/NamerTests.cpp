#include "doctest/doctest.h"
#include "ApprovalTests/namers/HelpMessages.h"
#include "ApprovalTests/namers/SeparateApprovedAndReceivedDirectoriesNamer.h"
#include "ApprovalTests/namers/NamerFactory.h"
#include "ApprovalTests/utilities/SystemUtils.h"
#include "../utilities/TestUtils.h"
#include <iostream>
#include <memory>

using namespace ApprovalTests;

TEST_CASE("AdditionalSections")
{
    auto namer = Approvals::getDefaultNamer();

    {
        auto section_namer = NamerFactory::appendToOutputFilename("case1");
        require_ends_with(namer->getApprovedFile(".txt"),
                          "NamerTests.AdditionalSections.case1.approved.txt");
    }
    {
        auto section_namer = NamerFactory::appendToOutputFilename("case2");
        require_ends_with(namer->getApprovedFile(".txt"),
                          "NamerTests.AdditionalSections.case2.approved.txt");
    }
    require_ends_with(namer->getApprovedFile(".txt"),
                      "NamerTests.AdditionalSections.approved.txt");
}

struct TestNameResetter
{
    TestNameResetter() : oldPrefix_(TestName::directoryPrefix)
    {
    }
    ~TestNameResetter()
    {
        TestName::directoryPrefix = oldPrefix_;
    }
    std::string oldPrefix_;
};

std::string getFileName()
{
    return "ApprovalTests.cpp/tests/DocTest_Tests/namers/NamerTests.cpp";
}

TEST_CASE("Find from parent")
{
    TestNameResetter resetter;
    TestName name;
    std::string junkDir = "/non/existing/directory/";
    TestName::directoryPrefix = junkDir;
    auto file = name.checkParentDirectoriesForFile(getFileName());
    //    std::cout << "Working directory = " << std::filesystem::absolute(".") << std::endl;
    std::cout << "File name = " << file << std::endl;
    CHECK(FileUtils::fileExists(file));
    CHECK(TestName::directoryPrefix != junkDir);
}

TEST_CASE("Check non-existent file reports misconfigured build")
{

    TestNameResetter resetter;
    TestName name;
    TestName::directoryPrefix = "";
    std::string nonExistentPath = "/I/Do/Not/Exist.cpp";
    name.setFileName(nonExistentPath);
    try
    {
        auto file = name.getFileName(); // Simulate no tests having yet been run
        std::cout << "File name = " << file << std::endl;
        FAIL("TestName::getFileName() on non-existent file should have thrown");
    }
    catch (const std::runtime_error& e)
    {
        std::string exceptionText = e.what();
        CHECK(StringUtils::contains(exceptionText, nonExistentPath));
        CHECK(StringUtils::contains(
            exceptionText, "There seems to be a problem with your build configuration."));
    }
}

TEST_CASE("Test Root Directory Help Message")
{
    Approvals::verify(HelpMessages::getUnconfiguredRootDirectory());
}
