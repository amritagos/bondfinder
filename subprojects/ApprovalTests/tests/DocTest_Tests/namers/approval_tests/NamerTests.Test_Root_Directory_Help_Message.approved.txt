*****************************************************************************
*                                                                           *
* Hello from Approval Tests.
*
* It looks like your program is calling some code that requires knowledge of
* the location of the main() in your source tree.
*
* To do this, add the following to your main.cpp file:
*
*     APPROVAL_TESTS_REGISTER_MAIN_DIRECTORY
*
* Currently, this is only required if you are using TemplatedCustomNamer's
* {RelativeTestSourceDirectory}.
*                                                                           *
*****************************************************************************
